src/hotspot/cpu/aarch64/register_aarch64.hpp

```diff
@@ -34,42 +34,44 @@ typedef VMRegImpl* VMReg;
 
 // Use Register as shortcut
 class RegisterImpl;
-typedef RegisterImpl* Register;
+typedef const RegisterImpl* Register;
 
-inline const Register as_Register(int encoding) {
-  return (Register)(intptr_t) encoding;
-}
+inline constexpr Register as_Register(int encoding);
 
 class RegisterImpl: public AbstractRegisterImpl {
- public:
+  static constexpr Register first();
+
+public:
   enum {
     number_of_registers         =   32,
     number_of_byte_registers      = 32,
-    number_of_registers_for_jvmci = 34,  // Including SP and ZR.
+    number_of_declared_registers  = 34,  // Including SP and ZR.
     max_slots_per_register = 2
   };
 
   // derived registers, offsets, and addresses
-  Register successor() const                          { return as_Register(encoding() + 1); }
+  const Register successor() const { return this + 1; }
 
   // construction
-  inline friend const Register as_Register(int encoding);
+  inline friend constexpr Register as_Register(int encoding);
 
-  VMReg as_VMReg();
+  VMReg as_VMReg() const;
 
   // accessors
-  int   encoding() const                         { assert(is_valid(), "invalid register"); return (intptr_t)this; }
-  bool  is_valid() const                         { return 0 <= (intptr_t)this && (intptr_t)this < number_of_registers; }
-  bool  has_byte_register() const                { return 0 <= (intptr_t)this && (intptr_t)this < number_of_byte_registers; }
+  int encoding() const             { assert(is_valid(), "invalid register"); return encoding_nocheck(); }
+  bool is_valid() const            { return this >= first() && this - first() < number_of_registers; }
```

Comment:
Some tiny suggestions, an unsigned comparison between `this - first()` and `number_of_registers` would be sufficient here.
```suggestion
  bool is_valid() const            { (unsigned)(this - first()) < number_of_registers; }
```

Reply:
OK.
To my surprise, this really does generate better code. I've been a GCC maintainer for a very long time, and I could have sworn that we did this optimization in the last century.